MapboxCoreNavigationTests/NavigationServiceTests.swift

```diff
@@ -617,6 +617,110 @@ class NavigationServiceTests: XCTestCase {
+            "Regent Street #1",
+            "Regent Street #2",
+            "Air Street",
+            "Brewer Street",
+            "Lexington Street",
+            "Broadwick Street #1",
+            "Broadwick Street #2",
+            "Dufour\'s Place",
+            "Dropped Pin #1"
+        ]
+        
+        if let route = route {
+            XCTAssertEqual(route.legs[0].steps.count, 11, "Unexpected number of steps.")
+            
+            let navigationService = MapboxNavigationService(route: route, routeOptions: routeOptions, simulating: .never)
+            let routeController = navigationService.router as! RouteController
+            
+            // Iterate over each step and simulate movement along the route.
+            route.legs[0].steps.enumerated().forEach {
+                let stepCoordinates = $0.element.shape!.coordinates
+                let stepLocations: [CLLocation] = stepCoordinates.enumerated().map {
+                    let currentCoordinate = $0.element
+                    return CLLocation(latitude: currentCoordinate.latitude, longitude: currentCoordinate.longitude)
+                }
+                
+                stepLocations.enumerated().forEach {
+                    let currentSpokenInstruction = navigationService.routeProgress.currentLegProgress.currentStepProgress.currentSpokenInstruction!.text
+                    XCTAssertFalse(voiceInstructionsArray.isEmpty, "Voice instructions array should not be empty.")
+                    if currentSpokenInstruction != voiceInstructionsArray.first {
+                        voiceInstructionsArray.removeFirst()
+                    }
+
+                    XCTAssertEqual(currentSpokenInstruction, voiceInstructionsArray.first)
+                    
+                    let currentBannerInstruction = navigationService.routeProgress.currentLegProgress.currentStepProgress.currentVisualInstruction!.primaryInstruction.text!
+                    XCTAssertFalse(bannerInstructionsArray.isEmpty, "Banner instructions array should not be empty.")
+                    if currentBannerInstruction != bannerInstructionsArray.first {
+                        bannerInstructionsArray.removeFirst()
+                    }
+                    
+                    XCTAssertEqual(currentBannerInstruction, bannerInstructionsArray.first)
+                    
+                    // Attempt to snarf instructions by calling methods which internally call Navigator.getStatusForTimestamp(_:) multiple times.
+                    let outOfBandCallsCount = 5
+                    for _ in 0..<outOfBandCallsCount {
+                        let _ = navigationService.router.userIsOnRoute($0.element)
+                        let _ = routeController.location
+                        let _ = routeController.snappedLocation
+                        navigationService.router!.locationManager!(navigationService.locationManager, didUpdateLocations: [$0.element])
```

Comment:
Why is location update called multiple times (inside the cycle)? Is it to simulate the case when users manually advance the progress?

Reply:
Yep, basically idea is to call methods which internally call `getStatusForTimestamp:` to see whether data it returns is correct and doesn't lead to any issues on platform side (like showing voice/visual instructions).